### PR TITLE
Issue-217: Clean the unnecessary ConnectionFactory creation

### LINKED ISSUE #217: Connection pooling
If a user is writing to or reading from multiple streams, there is no requirement for these to go over separate connections. As the number of streams grows this becomes more and more important.
For this reason we should implement connection pooling so that redundant connections to the same host are not needed.

### PR DESCRIPTION
**Pull request by huide9**
Signed-off-by: huide9 <huide9@gmail.com>

**Change log description**  
Clean the unnecessary ConnectionFactory instance creation; modify the integration test case to apply testing for the changed code. 

**Purpose of the change**  
A step towards resolving #217: connection pooling  - reduce the connection instance creation. 

**What the code does**  
Replace `new ConnectionFactoryImpl()` with input instance which created by StreamManager and other management class, such as EventStreamClientFactory.
*new instance creation removed from: 
- BatchClientFactory,
- ByteStreamClientFactory, 
- SynchronizerClientFactory 
- ReaderGroumanager. 
- The integration test case calls.

**How to verify it**  
integration tests - gradlew test:integration:build

### CODE REVIEW

File: client/src/main/java/io/pravega/client/ByteStreamClientFactory.java [hunk 1]
```diff
@@ -33,18 +32,19 @@
      *
      * @param scope The scope string.
      * @param config Configuration for the client.
+     * @param connectionFactory Connection for the client.
      * @return Instance of ByteStreamClientFactory implementation.
      */
-    static ByteStreamClientFactory withScope(String scope, ClientConfig config) {
-        val connectionFactory = new ConnectionFactoryImpl(config);
+    static ByteStreamClientFactory withScope(String scope, ClientConfig config, ConnectionFactory connectionFactory) {
+        /* val connectionFactory = new ConnectionFactoryImpl(config); */
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/ByteStreamClientFactory.java]: Do not commit commented out code. Just remove it.

File: test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/ClientReader.java [hunk 2]
```diff
@@ -151,11 +153,12 @@ private StreamReader getReader(String streamName) {
...
-            try (ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(ClientAdapterBase.SCOPE,
-                    ClientConfig.builder().controllerURI(ClientReader.this.controllerUri)
-                            .trustStore("../../config/cert.pem")
-                            .credentials(new DefaultCredentials("1111_aaaa", "admin"))
-                            .validateHostName(false).build())) {
+            ClientConfig clientConfig = ClientConfig.builder().controllerURI(ClientReader.this.controllerUri)
+                    .trustStore("../../config/cert.pem")
+                    .credentials(new DefaultCredentials("1111_aaaa", "admin"))
+                    .validateHostName(false).build();
+            ConnectionFactory connectionFactory = new ConnectionFactoryImpl(clientConfig);
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/ClientReader.java]: `ClientFactory` implements `AutoCloseable`. You need to keep a reference to this object and close it when appropriate

Comment (andreipaduroiu) [reply test/integration/src/main/java/io/pravega/test/integration/selftest/adapters/ClientReader.java]: Please fix this everywhere in your pull request. 

I have also seen instances of `ReaderGroupManager` not being closed when done with.

**Comment by huide9** [comment]
@andreipaduroiu The changes are mainly for input parameters for 4 classes and their test cases (integration / system); any idea about '**codecov/patch - 0% of diff hit**' here?  thanks a lot.

File: client/src/main/java/io/pravega/client/BatchClientFactory.java [hunk 3]
```diff
@@ -43,10 +42,10 @@
      *
      * @param scope The scope of the stream.
      * @param config Configuration for the client.
+     * @param connectionFactory Connection for the client.
      * @return Instance of BatchClientFactory implementation.
      */
-    static BatchClientFactory withScope(String scope, ClientConfig config) {
-        val connectionFactory = new ConnectionFactoryImpl(config);
+    static BatchClientFactory withScope(String scope, ClientConfig config, ConnectionFactory connectionFactory) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/BatchClientFactory.java]: This is a public interface and this change is backwards incompatible. We don't want users to have to add an additional step to get their client running.

File: client/src/main/java/io/pravega/client/ByteStreamClientFactory.java [hunk 4]
```diff
@@ -33,18 +32,18 @@
      *
      * @param scope The scope string.
      * @param config Configuration for the client.
+     * @param connectionFactory Connection for the client.
      * @return Instance of ByteStreamClientFactory implementation.
      */
-    static ByteStreamClientFactory withScope(String scope, ClientConfig config) {
-        val connectionFactory = new ConnectionFactoryImpl(config);
+    static ByteStreamClientFactory withScope(String scope, ClientConfig config, ConnectionFactory connectionFactory) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/ByteStreamClientFactory.java]: Again this is a public interface. We can't make this change.

File: client/src/main/java/io/pravega/client/SynchronizerClientFactory.java [hunk 5]
```diff
@@ -33,10 +32,10 @@
      *
      * @param scope The scope string.
      * @param config Configuration for the client.
+     * @param connectionFactory Connection for the client.
      * @return Instance of ClientFactory implementation.
      */
-    static SynchronizerClientFactory withScope(String scope, ClientConfig config) {
-        val connectionFactory = new ConnectionFactoryImpl(config);
+    static SynchronizerClientFactory withScope(String scope, ClientConfig config, ConnectionFactory connectionFactory) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/SynchronizerClientFactory.java]: Again this is a public interface. We can't make this change.

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java [hunk 6]
```diff
@@ -29,21 +30,23 @@
      *
      * @param scope The Scope string.
      * @param controllerUri The Controller URI.
+     * @param connectionFactory The Connection for the client.
      * @return Instance of Stream Manager implementation.
      */
-    public static ReaderGroupManager withScope(String scope, URI controllerUri) {
-        return withScope(scope, ClientConfig.builder().controllerURI(controllerUri).build());
+    public static ReaderGroupManager withScope(String scope, URI controllerUri, ConnectionFactory connectionFactory) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java]: Again this is a public interface. We can't make this change.

File: client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java [hunk 7]
```diff
@@ -29,21 +30,23 @@
...
      * Creates a new instance of ReaderGroupManager.
      *
      * @param scope The Scope string.
      * @param clientConfig Configuration for the client.
+     * @param connectionFactory The Connection for the client.
      * @return Instance of Stream Manager implementation.
      */
-    public static ReaderGroupManager withScope(String scope, ClientConfig clientConfig) {
-        return new ReaderGroupManagerImpl(scope, clientConfig, new ConnectionFactoryImpl(clientConfig));
+    public static ReaderGroupManager withScope(String scope, ClientConfig clientConfig, ConnectionFactory connectionFactory) {
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java]: Again this is a public interface. We can't make this change.

**Comment by andreipaduroiu** [comment]
@huide9 @tkaitchuck What is the status of this? I see no activity in 7 days. Are we still pursuing this change?

**Comment by huide9** [comment]
Close as the solution is not proposed one. Another PR should be opened for review later.

**Pull request closed without merging**